### PR TITLE
SageMaker controller machine learning model training tutorial

### PR DESCRIPTION
**Pull request by judyheflin**
Description of changes: I added a Tutorials section to the docs and included a tutorial on how to train a machine learning model with the SageMaker ACK service controller using Amazon EKS. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+For more information on authorization and access for ACK service controllers, including details regarding recommended IAM policies, see [Configure Permissions][configure-permissions].
+
+### Install the SageMaker ACK service controller
+
+Get the SageMaker Helm chart and make it available on the `Deployment` host with the following commands:
+
+```bash
+export HELM_EXPERIMENTAL_OCI=1
+export SERVICE=sagemaker
+export RELEASE_VERSION=v0.0.4
```
Comment (jaypipes) [inline docs/content/docs/tutorials/sagemaker-example.md:121]: Do we want to put a specific release version here? We might do something like this instead?

```
LATEST_RELEASE_VERSION=`curl -sL https://api.github.com/repos/aws-controllers-k8s/sagemaker-controller/releases/latest | grep '"tag_name":' | cut -d'"' -f4`
RELEASE_VERSION=${LATEST_RELEASE_VERSION:-v0.0.4}
```

Once Sagemaker releases a non-pre-prod release of sagemaker-controller, then the `RELEASE_VERSION` variable will contain that latest release version, otherwise it will be v0.0.4...

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+```bash
+printf '
+#!/usr/bin/env bash
+# Create the S3 bucket
+if [[ $SERVICE_REGION != "us-east-1" ]]; then
+  aws s3api create-bucket --bucket "$SAGEMAKER_BUCKET" --region "$SERVICE_REGION" --create-bucket-configuration LocationConstraint="$SERVICE_REGION"
+else
+  aws s3api create-bucket --bucket "$SAGEMAKER_BUCKET" --region "$SERVICE_REGION"
+' > ./create-bucket.sh
+```
```
Comment (jaypipes) [inline docs/content/docs/tutorials/sagemaker-example.md:188]: certainly not required for this PR, but it would be awesome in the future to have a tutorial that used the S3 ACK controller and an [S3 Bucket CR](https://aws-controllers-k8s.github.io/community/reference/s3/v1alpha1/bucket/) instead of calling the `aws s3api create-bucket` command :)

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+Specify your region-specific XGBoost image URI: 
+
+```bash
+export XGBOOST_IMAGE=683313688378.dkr.ecr.us-east-1.amazonaws.com/sagemaker-xgboost:1.2-1
+```
+{{% hint type="info" title="Change XGBoost image URI based on region" %}}
+**IMPORTANT**: If your `SERVICE_REGION` is not `us-east-1`, you must change the `XGBOOST_IMAGE` URI. To find your region-specific XGBoost image URI, choose your region in the [SageMaker Docker Registry Paths page](https://docs.aws.amazon.com/sagemaker/latest/dg/sagemaker-algo-docker-registry-paths.html), and then select **XGBoost (algorithm)**. For this example, use version 1.2-1.
+{{% /hint %}}
+
+Next, create a `training.yaml` file to specify the parameters for your SageMaker training job. This file specifies your SageMaker training job name, any relevant hyperperameters, and the location of your training and validation data. You can also use this document to specify which Amazon Elastic Container Registry (ECR) image to use for training.
```
Comment (jaypipes) [inline docs/content/docs/tutorials/sagemaker-example.md:237]: "hyperperameters" should be "hyperparameters"

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+    max_depth: "5"
+    gamma: "4"
+    eta: "0.2"
+    min_child_weight: "6"
+    objective: "multi:softmax"
+    num_class: "10"
+    num_round: "10"
+  algorithmSpecification:
+    # The URL and tag of your ECR container
+    # If you are not on us-west-1 you can find an imageURI here https://docs.aws.amazon.com/sagemaker/latest/dg/sagemaker-algo-docker-registry-paths.html
```
Comment (jaypipes) [inline docs/content/docs/tutorials/sagemaker-example.md:258]: Note that the region specified above as SERVICE_REGION is us-east-1, no us-west-1. I'd just remove the above comment line, actually.

**Review by jaypipes — CHANGES_REQUESTED** [review]
@judyheflin great work on this! A few relatively minor suggestions inline for you :)

File: docs/content/docs/tutorials/sagemaker-example.md (hunk 2 above)
Comment (judyheflin) [reply docs/content/docs/tutorials/sagemaker-example.md:188]: I agree! Adding to To-do's :)

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+## Train an XGBoost model
+
+### Prepare your data
+
+For training a model with SageMaker, we will need an S3 bucket to store the dataset and model training artifacts. For this example, we will use [MNIST](http://yann.lecun.com/exdb/mnist/) data stored in [LIBSVM](https://www.csie.ntu.edu.tw/~cjlin/libsvm/) format.
+
+First, create a variable for the S3 bucket:
+
+```bash
+export RANDOM_VAR=$RANDOM
```
Comment (RedbackThomson) [inline docs/content/docs/tutorials/sagemaker-example.md:163]: I'd recommend you use the account ID:
```bash
export ACCOUNT_ID=$(aws sts get-caller-identity --query "Account" --output text)
```

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+aws iam delete-role --role-name $SAGEMAKER_EXECUTION_ROLE_NAME
+
+# Delete IAM role created for IRSA
+aws iam detach-role-policy --role-name $OIDC_ROLE_NAME --policy-arn arn:aws:iam::aws:policy/AmazonSageMakerFullAccess
+aws iam delete-role --role-name $OIDC_ROLE_NAME
+```
+
+To delete your EKS clusters, pods, or nodes, see [Amazon EKS Setup - Cleanup][cleanup-eks].  
+
+[configure-permissions]: https://aws-controllers-k8s.github.io/community/docs/user-docs/authorization/
```
[the remark this answers is not in the document]
Comment (RedbackThomson) [reply docs/content/docs/tutorials/sagemaker-example.md]: I've been using relative paths for the moment, in case we change root URLs in the future. 

Also the `/docs` is a Hugo setting, because we are viewing the "documentation" part of the website - as opposed to the API reference part. That is a fluid property that's open to discussion, but it's essentially because of our file structure.

**Review by RedbackThomson — COMMENTED** [review]
Looks very clean. Had a few nits in the comments.

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+toc: true
+---
+
+The SageMaker ACK service controller makes it easier for machine learning developers and data scientists who use Kubernetes as their control plane to train, tune, and deploy machine learning models in Amazon SageMaker without logging into the SageMaker console. 
+
+The following steps will guide you through the setup and use of the Amazon SageMaker ACK service controller for training a machine learning model.
+
+## Setup
+
+Although it is not necessary to use Amazon Elastic Kubernetes Service (Amazon EKS) with ACK, this guide assumes that you have access to an Amazon EKS cluster. For automated cluster creation using `eksctl`, see [Getting started with Amazon EKS - `eksctl`](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-eksctl.html) and create your cluster with Amazon EC2 Linux managed nodes. If this is your first time creating an Amazon EKS cluster, see [Amazon EKS Setup](https://docs.aws.amazon.com/deep-learning-containers/latest/devguide/deep-learning-containers-eks-setup.html).
```
[the remark this answers is not in the document]
Comment (judyheflin) [reply docs/content/docs/tutorials/sagemaker-example.md:19]: Some people on the team had trouble starting up their EKS clusters when we ran through this and found that the Linux managed nodes steps worked because the other eksctl steps didn't create nodes.

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+    - [yq](https://mikefarah.gitbook.io/yq) - A command line tool for YAML processing.
+      - Linux
+        ```
+        sudo wget https://github.com/mikefarah/yq/releases/download/v4.9.8/yq_linux_amd64 -O /usr/bin/yq
+        sudo chmod +x /usr/bin/yq
+        ```
+      - Mac
+        ```
+        brew install yq
+        ```
```
[the remark this answers is not in the document]
Comment (surajkota) [reply docs/content/docs/tutorials/sagemaker-example.md]: @jaypipes yes. `yq` tool works properly if installed using the binaries on Linux. using snap it doesn't work

@judyheflin please add a note here saying install using binaries on Linux and point to this link - https://mikefarah.gitbook.io/yq/#wget

[the remark this answers is not in the document]
Comment (jaypipes) [reply docs/content/docs/tutorials/sagemaker-example.md]: @surajkota @judyheflin please let's just do this:

```
- [yq](https://mikefarah.gitbook.io/yq/#wget) - A command line tool for YAML processing.
```

and no Linux/Mac installation instructions here.

[the remark this answers is not in the document]
Comment (judyheflin) [reply docs/content/docs/tutorials/sagemaker-example.md]: Added!

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+
+This guide assumes that you have:
+  - Created an EKS cluster with Kubernetes version 1.16 or higher. 
+  - AWS IAM permissions to create roles and attach policies to roles.
+  - Installed the following tools on the client machine used to access your Kubernetes cluster:
+    - [kubectl](https://docs.aws.amazon.com/eks/latest/userguide/install-kubectl.html) - A command line tool for working with Kubernetes clusters. 
+    - [Helm](https://helm.sh/docs/intro/install/) - A tool for installing and managing Kubernetes applications.
+    - [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv1.html) - A command line tool for interacting with AWS services. 
+    - [eksctl](https://docs.aws.amazon.com/eks/latest/userguide/eksctl.html) - A command line tool for working with EKS clusters.
+    - [yq](https://mikefarah.gitbook.io/yq) - A command line tool for YAML processing. (For Linux installation, use the [`wget` plain binary installation](https://mikefarah.gitbook.io/yq/#wget))
```
Comment (jaypipes) [inline docs/content/docs/tutorials/sagemaker-example.md:31]: OK, I guess this is fine, too.

**Review by jaypipes — APPROVED** [review]
:+1:

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+
+```bash
+kubectl apply -f $CHART_EXPORT_PATH/$SERVICE-chart/crds
+```
+
+Create a namespace and install the SageMaker ACK service controller with the Helm chart. 
+```bash
+helm install -n $ACK_K8S_NAMESPACE --create-namespace --skip-crds ack-$SERVICE-controller \
+ $CHART_EXPORT_PATH/$SERVICE-chart
+```
```
[the remark this answers is not in the document]
Comment (surajkota) [reply docs/content/docs/tutorials/sagemaker-example.md]: Suggest leaving it this way. It was discussed in a separate thread that Helm shouldn't be used to manage CRDs. CRDs don't get removed with the helm uninstall command. I also had an instance where CRD did not get updated if it already exists.

apply crd
install controller
uninstall controller
remove crd

I think we should recommend this in main install doc as well

File: docs/content/docs/tutorials/sagemaker-example.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (surajkota) [reply docs/content/docs/tutorials/sagemaker-example.md]: Yes, the main creating section has command to [create without-nodegroup](https://docs.aws.amazon.com/eks/latest/userguide/create-cluster.html). so people have always missed adding worker nodes. this is the only other simple set of commands we found to create a cluster with worker nodes

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+### Prepare your data
+
+For training a model with SageMaker, we will need an S3 bucket to store the dataset and model training artifacts. For this example, we will use [MNIST](http://yann.lecun.com/exdb/mnist/) data stored in [LIBSVM](https://www.csie.ntu.edu.tw/~cjlin/libsvm/) format.
+
+First, create a variable for the S3 bucket:
+
+```bash
+export ACCOUNT_ID=$(aws sts get-caller-identity --query "Account" --output text)
+# Specify your service region
+export SERVICE_REGION=us-east-1
```
Comment (surajkota) [inline docs/content/docs/tutorials/sagemaker-example.md:165]: the default region for the controller is defined here - https://github.com/aws-controllers-k8s/community/pull/945/files#diff-785fe064cc3ba92455af3ccd1dd44ddd6714e1b9cc6c2a11b3bbdd12fdae4cd5R131.

If the region for the job/resource is different from the default controller region, we will need this - https://github.com/aws-controllers-k8s/sagemaker-controller#50-cross-region-resource-management

For this tutorial, recommend keeping everything in one region or change [the substituted region here](https://github.com/aws-controllers-k8s/community/pull/945/files#diff-785fe064cc3ba92455af3ccd1dd44ddd6714e1b9cc6c2a11b3bbdd12fdae4cd5R131) to `SERVICE_REGION`

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+
+Install the SageMaker ACK service controller and relevant custom resource definitions (CRDs) with the Helm chart. 
+```bash
+helm install -n $ACK_K8S_NAMESPACE --create-namespace ack-$SERVICE-controller \
+ $CHART_EXPORT_PATH/$SERVICE-chart
+```
+
+Verify that the CRDs and Helm charts were deployed with the following commands:
+```bash
+kubectl get crds
```
Comment (surajkota) [inline docs/content/docs/tutorials/sagemaker-example.md:150]: nit `kubectl get crds | grep "services.k8s.aws"`

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+
+```bash
+NAME                            SECONDARYSTATUS   TRAININGJOBSTATUS
+ack-xgboost-training-job-7420   Starting          InProgress
+ack-xgboost-training-job-7420   Downloading       InProgress
+ack-xgboost-training-job-7420   Training          InProgress
+ack-xgboost-training-job-7420   Completed         Completed
+```
+
+If your training job failed, look to the `status.failureReason` after running the following command:
```
Comment (surajkota) [inline docs/content/docs/tutorials/sagemaker-example.md:313]: lets talk about success in addition to failure case here. say to user something like,

to see the details about the training job, run `describe ....` if `status.trainingJobStatus` is `Completed`, find the model location under `status.modelArtifacts.s3ModelArtifacts`. If `status.trainingJobStatus` is `Failed` look for `status.failureReason`

File: docs/content/docs/tutorials/sagemaker-example.md [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+### Cleanup
+
+You can delete your SageMaker training job with the `kubectl delete` command.
+```bash
+kubectl delete trainingjob $JOB_NAME
+```
+
+To remove the SageMaker ACK service controller, related CRDs, and namespaces see [ACK Cleanup][cleanup].
+
+It is recommended to delete any additional resources such as S3 buckets, IAM roles, and IAM policies when you no longer need them. You can delete these resources with the following commands or directly in the AWS console.
```
Comment (surajkota) [inline docs/content/docs/tutorials/sagemaker-example.md:331]: ++, Nice!!

**Review by surajkota — CHANGES_REQUESTED** [review]
Nice work!! few comments

File: docs/content/docs/tutorials/sagemaker-example.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (judyheflin) [reply docs/content/docs/tutorials/sagemaker-example.md]: Leaving this as is.

**Review by RedbackThomson — APPROVED** [review]
Thanks for the updates. Will leave it to Suraj for final changes, then I can merge.
